README.md
# cheryl_birthday

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-11-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-11-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/cheryl_birthday/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/cheryl_birthday/compare/0.1.0...0.1.1


test/cheryl_birthday/core_test.clj
(ns cheryl-birthday.core-test
  (:require [clojure.test :refer :all]
            [cheryl-birthday.core :refer :all]
            [clojure.string :as s]))

(def dates ["May 15", "May 16", "May 19", "June 17", "June 18", "July 14", "July 16", "August 14", "August 15", "August 17"])

(deftest test-know?
  (testing "having muliple options is not knowing"
    (is (not (know? ["May 3" "May 6"]))))

  (testing "having no option at all is not knowing"
    (is (not (know? []))))

  (testing "having exactly one option is knowing"
    (is (know? ["May 6"]))))

(deftest test-tell
  (testing "telling the month narrows down possible answers to that month"
    (is (= ["May 15" "May 16" "May 19"] (tell "May" dates))))
  (testing "telling using month"
    (is (= ["May 15" "May 16" "May 19"] (tell (month "May 15") dates))))
  (testing "telling the day narrows down possible answers to that day"
    (is (= ["May 15" "August 15"] (tell "15" dates))))
  (testing "telling using day"
    (is (= ["May 15" "August 15"] (tell (day "May 15") dates)))))

(deftest test-albert-doesn't-know-and-he-knows-bernard-can't-tell-either
  (testing "Albert doesn't know if there are multiple dates in that month"
    (is (albert-doesn't-know-and-he-knows-bernard-can't-tell-either "August 15" dates)))
  (testing "Bernard can't tell if there are multiple dates on that day"
    (is (albert-doesn't-know-and-he-knows-bernard-can't-tell-either "August 15" dates)))
  (testing "it cannot be May because otherwise Bernard might have a unique answer in May 19"
    (is (not (albert-doesn't-know-and-he-knows-bernard-can't-tell-either "May 15" dates))))
  (testing "it cannot be June because otherwise Bernard might have a unique answer in June 18"
    (is (not (albert-doesn't-know-and-he-knows-bernard-can't-tell-either "June 17" dates)))))

(deftest test-bernard-couldn't-tell-at-first-but-now-he-knows
  (testing "The day part cannot be unique for otherwise Bernard can tell"
    (is (not (bernard-couldn't-tell-at-first-but-now-he-knows "May 19" dates))))
  (testing "Bernard knows Albert wasn't told May or June. Among dates in July and August, the day part must be unique"
    (is (not (bernard-couldn't-tell-at-first-but-now-he-knows "July 14" dates))))
  (testing "Bernard knows Albert wasn't told May or June. Among dates in July and August, the day part must be unique"
    (is (not (bernard-couldn't-tell-at-first-but-now-he-knows "August 14" dates))))
  (testing "Bernard knows Albert wasn't told May or June. Among dates in July and August, the day part must be unique"
    (is (bernard-couldn't-tell-at-first-but-now-he-knows "July 16" dates)))
  (testing "Bernard knows Albert wasn't told May or June. Among dates in July and August, the day part must be unique"
    (is (bernard-couldn't-tell-at-first-but-now-he-knows "August 15" dates)))
  (testing "Bernard knows Albert wasn't told May or June. Among dates in July and August, the day part must be unique"
    (is (bernard-couldn't-tell-at-first-but-now-he-knows "August 17" dates))))

(deftest test-albert-finally-finds-out-too
  (testing "The month part of the remaining dates must be unique for Albert to know"
    (is (then-i-also-know-when-cheryl's-birthday-is "July 16" ["July 16" "August 15" "August 17"])))
  (testing "Albert can't tell if the month part isn't unique"
    (is (not (then-i-also-know-when-cheryl's-birthday-is "August 15" ["July 16" "August 15" "August 17"])))))

(run-tests)


src/cheryl_birthday/core.clj
(ns cheryl-birthday.core
  (:require [clojure.string :as s]))

(defn know? [dates]
  (= 1 (count dates)))

(defn tell [part dates]
  (filter #(s/includes? % part) dates))

(defn month [date]
  (first (s/split date #" ")))

(defn day [date]
  (second (s/split date #" ")))

(defn albert-doesn't-know-and-he-knows-bernard-can't-tell-either [date all-dates]
  (let [what-albert-narrows-down-to (tell (month date) all-dates)
        all-days-in-albert's-list (distinct (map day what-albert-narrows-down-to))
        what-albert-thinks-bernard-can-narrow-down-to (map #(tell % all-dates) all-days-in-albert's-list)
        bernard-can't-identify-a-unique-date (not-any? know? what-albert-thinks-bernard-can-narrow-down-to)]
    (and (not (know? what-albert-narrows-down-to)) bernard-can't-identify-a-unique-date)))

(defn bernard-couldn't-tell-at-first-but-now-he-knows [date all-dates]
  (let [what-bernard-narrows-down-to-at-first (tell (day date) all-dates)
        what-bernard-narrows-down-to-after-hearing-albert (filter #(albert-doesn't-know-and-he-knows-bernard-can't-tell-either % all-dates) what-bernard-narrows-down-to-at-first)]
    (and (not (know? what-bernard-narrows-down-to-at-first)) (know? what-bernard-narrows-down-to-after-hearing-albert))))

(defn then-i-also-know-when-cheryl's-birthday-is [date all-dates]
  (let [what-albert-further-narrows-down-to (tell (month date) all-dates)]
    (know? what-albert-further-narrows-down-to)))


cheryl.clj
(require '[clojure.string :as str])

(def possible-dates '("May 15", "May 16", "May 19",
                          "June 17", "June 18",
                          "July 14", "July 16",
                          "August 14", "August 15", "August 17"
                    )
)

(defn tell [part]
  (filter #(.contains % part) possible-dates)
)

(defn know [dates]
  (= 1 (count dates))
)

(defn month [date]
  (first (str/split date #" "))
)

(defn day [date]
  (second (str/split date #" "))
)

(defn statement3 [date]
  (let [candidates (tell (month date))]
    (and
      (not (know candidates))
      (not-any? true? (map know (map tell (map day candidates))))
    )
  )
)

(defn statement4 [date]
  (let [at-first (tell (day date))]
    (and
      (not (know at-first))
      (know (filter statement3 at-first))
    )
  )
)

(defn statement5 [date]
  (know (filter statement4 (tell (month date))))
)

(defn statements3to5 [date]
  (and (statement3 date) (statement4 date) (statement5 date))
)

(def cheryl-birthday
  (filter statements3to5 possible-dates)
)

cheryl-birthday

(know cheryl-birthday)










